### PR TITLE
`python pin` now finds workspace root folder when called in subfolder

### LINKED ISSUE #4971: `uv python pin` should find `.python-version` file in a workspace
If working in a child directory of a workspace, `uv python pin` should support finding the `.python-version` file in the workspace root.

Related #4970

### PR DESCRIPTION
**Pull request by silvanocerza**
## Summary

Fixes #4971.

This PR changes `python pin` behaviour when called in a workspace subfolder. Previously it would only read or create a `.python-version` file in the current working directory, now instead it will first search for the workspace root folder and read or create the `.python-version` file in that folder. 

We assume the root folder contains a valid `pyproject.toml` file.

If we're not in a workspace use the previous behaviour and creates or reads `.python-version` in the current working directory.

## Test Plan

I added a new test to verify this new behaviour, and run all tests locally.

### CODE REVIEW

**Comment by zanieb** [comment]
We might want to make this asymmetric per https://github.com/astral-sh/uv/issues/4971#issuecomment-2226958546 and _read_ from the workspace root by default but when _writing_ we should require an opt-in via a `--workspace` otherwise we should write to the current directory. Or should we use `--isolated` to ignore the workspace and write to the working directory? What do you think? I'm not 100% certain what the best UX is yet.

**Comment by T-256** [comment]
> Or should we use `--isolated` to ignore the workspace and write to the working directory?

IMO this is more correct to me, since pinned python version is usually used for venv creation and venv is in workspace root.

**Comment by silvanocerza** [comment]
I would agree with @T-256 here, it makes sense to me using the `--isolated` option to ignore the workspace. 
Should I got with that?

**Comment by silvanocerza** [comment]
Rebased to bring in CI fix from #5034 and remove some commented out code that I missed.

File: crates/uv/tests/python_pin.rs [hunk 1]
```diff
@@ -211,6 +213,44 @@ fn python_pin() {
...
+    // Create a subfolder
+    let working_dir = context.temp_dir.child("subfolder");
+    working_dir.create_dir_all().unwrap();
+
+    // Run python pin command in subfolder
+    uv_snapshot!(context.filters(), context.python_pin().current_dir(&working_dir).arg("3.11"), @r###"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+    Pinned to `3.11`
```
Comment (zanieb) [inline crates/uv/tests/python_pin.rs:242]: Can we add a message noting where we pinned when it differs from the pwd?

File: crates/uv/src/commands/python/pin.rs [hunk 2]
```diff
@@ -29,9 +29,16 @@ pub(crate) async fn pin(
         warn_user_once!("`uv python pin` is experimental and may change without warning.");
     }
 
+    let working_dir =
```
Comment (zanieb) [inline crates/uv/src/commands/python/pin.rs:32]: We may want to call this `target_dir` or something to avoid confusion with the current working directory.

File: crates/uv-python/src/version_files.rs [hunk 3]
```diff
@@ -32,7 +102,7 @@ pub async fn requests_from_version_file() -> Result<Option<Vec<PythonRequest>>,
 /// Read a [`PythonRequest`] from a version file, if present.
 ///
 /// Prefers `.python-version` then the first entry of `.python-versions`.
-/// If multiple Python versions are desired, use [`requests_from_version_files`] instead.
+/// If multiple Python versions are desired, use [`requests_from_version_file`] instead.
```
Comment (zanieb) [inline crates/uv-python/src/version_files.rs:105]: Why this change?

File: crates/uv-python/src/version_files.rs [hunk 4]
```diff
@@ -13,7 +13,77 @@ pub static PYTHON_VERSIONS_FILENAME: &str = ".python-versions";
...
+        .await?
+        .lines()
+        .map(ToString::to_string)
+        .collect();
+    Ok(Some(lines))
+}
+
+async fn read_version_file_in(root: &PathBuf) -> Result<Option<String>, io::Error> {
+    let mut version_file = PathBuf::from(root);
+    version_file.push(PYTHON_VERSION_FILENAME);
```
Comment (zanieb) [inline crates/uv-python/src/version_files.rs:71]: Should we just use `Path::join` here?

File: crates/uv-python/src/version_files.rs [hunk 5]
```diff
@@ -13,7 +13,77 @@ pub static PYTHON_VERSIONS_FILENAME: &str = ".python-versions";
...
+    debug!("Reading requests from `{}`", version_file.display());
+    let lines: Vec<String> = fs::tokio::read_to_string(version_file)
+        .await?
+        .lines()
+        .map(ToString::to_string)
+        .collect();
+    Ok(Some(lines))
+}
+
+async fn read_version_file_in(root: &PathBuf) -> Result<Option<String>, io::Error> {
```
Comment (zanieb) [inline crates/uv-python/src/version_files.rs:69]: This can be `impl AsRef<Path>` or `&Path` instead of `&PathBuf`

File: crates/uv-python/src/version_files.rs (hunk 3 above)
Comment (zanieb) [reply crates/uv-python/src/version_files.rs:105]: Is this supposed to be `requests_from_versions_file`?

File: crates/uv-python/src/version_files.rs [hunk 6]
```diff
@@ -13,7 +13,77 @@ pub static PYTHON_VERSIONS_FILENAME: &str = ".python-versions";
...
+            .into_iter()
+            .next()
+            .inspect(|_| debug!("Using the first version from `.python-versions`"))
+            .map(|version| PythonRequest::parse(&version)))
+    } else {
+        Ok(None)
+    }
+}
+
+async fn read_versions_file_in(root: &PathBuf) -> Result<Option<Vec<String>>, io::Error> {
```
Comment (zanieb) [inline crates/uv-python/src/version_files.rs:54]: Can you add some simple docstrings to these?

**Comment by zanieb** [comment]
Sure let's just make `--isolated` ignore the workspace to start and we can tackle more complex behavior separately?

File: crates/uv-python/src/version_files.rs (hunk 3 above)
Comment (silvanocerza) [reply crates/uv-python/src/version_files.rs:105]: Yes, the docstring was wrong, `requests_from_version_files` doesn't exist.

**Comment by silvanocerza** [comment]
Ready for review again. 👍

File: crates/uv/src/commands/python/pin.rs [hunk 7]
```diff
@@ -69,16 +78,24 @@ pub(crate) async fn pin(
...
+        format!("Pinned to `{output}`")
+    };
+
+    if target_dir != std::env::current_dir()? {
+        // Print the version file use to pin only
+        // if it's not in the current working directory
+        message = format!("{message} in `{}`", version_file.display());
+    };
+
+    writeln!(printer.stdout(), "{message}")?;
```
Comment (silvanocerza) [inline crates/uv/src/commands/python/pin.rs:98]: I'm not sure about this part, would something like this be better? 🤔 

```
    let message = if exists {
        format!("Replaced existing pin with `{output}`")
    } else {
        format!("Pinned to `{output}`")
    };

    let message = if target_dir != std::env::current_dir()? {
        // Print the version file use to pin only
        // if it's not in the current working directory
        format!("{message} in `{}`", version_file.display())
    } else {
        message
    };
```

**Comment by zanieb** [comment]
Going to be lots of conflicts with #5215 and https://github.com/astral-sh/uv/pull/4989 — I'll own resolving them.

**Comment by zanieb** [comment]
Okay I resolved the conflicts but then realized we need to make a lot of other changes here, e.g. every other command needs to respect pins in the workspace. Kind of complicated, I need to think about it before moving forward.

**Comment by zanieb** [comment]
Here's where I left off https://github.com/astral-sh/uv/compare/zb/pin-find

**Comment by silvanocerza** [comment]
If you can give me a list of the commands that actually needs to respect the pin I should be able to handle that, I'm just unsure about all the commands.

Also probably we need to isolate the logic that finds the pin location, it would simplify implementing it for the other commands.

**Comment by zanieb** [comment]
I think we need to walk up the directory tree looking for a version file like we do with `pyproject.toml` files for project roots. Then, every command that needs a Python interpreter needs to respect the version file when determining the default. We don't need to do it all at once, I think outside the preview commands only `venv` supports version files right now and it's gated by preview.

We may or may not want to stop searching at a boundary where we find a `pyproject.toml`. If we don't, then we need to check if the version is pinned to something incompatible with the current project's Python requirement. If the pin comes from a directory above the project, then we probably need to warn and use the `Requires-Python` value instead. 

It seems simplest to do something like:

- Add a utility to the version file module that finds the nearest version file. We probably want a struct for this so we can store the path as state and allow read / write.
- Add a utility to the workspace / project abstractions to reconcile the discovered version file with the project, e.g. we can check if it's location is a prefix of the project or if it's inside the project and check compatibility with the project Python requirement.
- Use in `uv venv` under preview mode
- Use in all the new preview commands where we currently respect version files
- Add a tracking issue enumerating the remaining commands

How does that sound? Kind of figuring this out as I go :)

**Comment by zanieb** [comment]
I'm taking this on in https://github.com/astral-sh/uv/pull/6370
